src/utils/handleRequest.ts

```diff
@@ -0,0 +1,111 @@
+import { StrictEventEmitter } from 'strict-event-emitter'
+import { MockedRequest, RequestHandler } from '../handlers/RequestHandler'
+import { ServerLifecycleEventsMap } from '../node/glossary'
+import { MockedResponse } from '../response'
+import { SharedOptions } from '../sharedOptions'
+import { ResponseLookupResult, getResponse } from './getResponse'
+import { onUnhandledRequest } from './request/onUnhandledRequest'
+import { readResponseCookies } from './request/readResponseCookies'
+
+export interface HandleRequestOptions<ResponseType> {
+  /**
+   * Transforms a `MockedResponse` instance returned from a handler
+   * to a response instance
+   */
+  transformResponse?(response: MockedResponse<string>): ResponseType
+
+  /**
+   * Invoked whenever returning a bypassed (as-is) response.
+   */
+  onBypassResponse?(request: MockedRequest): void
+
+  /**
+   * Invoked when the mocked response is ready to be sent.
+   */
+  onMockedResponse?(
+    response: ResponseType,
+    handler: DeepRequired<ResponseLookupResult>,
+  ): void
+
+  /**
+   * Invoked when the mocked response is sent.
+   * Respects the response delay duration.
+   */
+  onMockedResponseSent?(
+    response: ResponseType,
+    handler: DeepRequired<ResponseLookupResult>,
+  ): void
+}
+
+export async function handleRequest<
+  ResponseType extends Record<string, any> = MockedResponse<string>
+>(
+  request: MockedRequest,
+  handlers: RequestHandler[],
+  options: SharedOptions,
+  emitter: StrictEventEmitter<ServerLifecycleEventsMap>,
+  handleRequestOptions?: HandleRequestOptions<ResponseType>,
+): Promise<ResponseType | undefined> {
+  emitter.emit('request:start', request)
```

Comment:
Request cookies _must not_ be set here, as they are already set in `parseIsomorpchicRequest`.

If cookies are set here once more, they will get overridden in Node.js, as request cookies in Node.js are resolved to `{}` (no location to check if the cookies should be inherited).